ext/include/opentelemetry/ext/zpages/tracez_data_aggregator.h

```diff
@@ -0,0 +1,207 @@
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <condition_variable>
+#include <list>
+#include <map>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <unordered_set>
+#include <vector>
+
+#include "opentelemetry/ext/zpages/latency_boundaries.h"
+#include "opentelemetry/ext/zpages/tracez_processor.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+#include "opentelemetry/trace/canonical_code.h"
+
+using opentelemetry::sdk::trace::SpanData;
+using opentelemetry::trace::CanonicalCode;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace ext {
+namespace zpages {
+
+/**
+ * kMaxNumberOfSampleSpans is the maximum number of running, completed or error
+ * sample spans stored at any given time for a given span name.
+ * This limit is introduced to reduce memory usage by trimming sample spans
+ * stored.
+ */
+const int kMaxNumberOfSampleSpans = 5;
+
+/**
+ * TracezData is the span data to be displayed for tracez zpages that is
+ * stored for each span name.
+ */
+struct TracezData {
+  /**
+   * TODO: At this time the maximum count is unknown but a larger data type
+   * might have to be used in the future to store these counts to avoid overflow
+   */
+  unsigned int running_span_count;
+  unsigned int error_span_count;
+
+  /**
+   * completed_span_count_per_latency_bucket is an array that stores the count
+   * of spans for each of the 9 latency buckets.
+   */
+  std::array<unsigned int, kLatencyBoundaries.size()>
+      completed_span_count_per_latency_bucket;
+
+  /**
+   * sample_latency_spans is an array of lists, each index of the array
+   * corresponds to a latency boundary(of which there are 9).
+   * The list in each index stores the sample spans for that latency boundary.
+   */
+  std::array<std::list<std::shared_ptr<SpanData>>, kLatencyBoundaries.size()>
+      sample_latency_spans;
+
+  /**
+   * sample_error_spans is a list that stores the error samples for a span name.
+   */
+  std::list<std::shared_ptr<SpanData>> sample_error_spans;
+
+  /**
+   * sample_running_spans is a list that stores the running span samples for a
+   * span name.
+   */
+  std::list<SpanData*> sample_running_spans;
```

Comment:
Curious, which piece of code owns the SpanData, how is the data lifecycle managed?

Reply:
SpanData that you see for running spans is owned by Span and for completed and error spans is owned by the data aggregator. I've been thinking of using something more consistent like a struct of my own that copies the required data over. It's quite confusing at the moment

Reply:
Thanks for the clarification.
I was thinking if `sample_running_spans` would always have the running spans (that haven't been reclaimed) or there could be race condition (e.g. span data got reclaimed and we ended up dangling pointers).

Reply:
Yeah this is a valid concern, to be honest it isn't great working with pointers to unique pointers. I was thinking of copying the spandata into a struct when I get them.  Another unrelated thing that I am worried about is thread safety issues regarding the reading information from the spandata, a thread safe recordable can be implemented for this.
These two solutions together require a duplication of data that is in SpanData and I was hesitant to do this.

But to answer your question directly, I have a test RemovalOfRunningSpanWhenCompleted and I try to access running_span sample information after the span has ended before getting the aggregated data and it lets me do it so my best guess isn't any dangling pointers after Span has been destroyed. Let me know if that answers your question